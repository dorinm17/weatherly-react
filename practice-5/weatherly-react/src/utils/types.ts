import { createContext } from "react";

interface CityCoordinates {
  name: string;
  lat: number;
  lon: number;
  local_names: Record<string, string>;
  // In the future, I want to implement a feature where the user gets suggestions in the search bar
  country: string;
  state?: string;
}

type WeatherInfo = {
  dt: number;
  weather: Record<string, string | number>[];
  clouds: Record<string, number>;
  pop: number;
};

type CityInfo = Record<string, string | number | Record<string, number>>;

// Generic type for forecast lists
type ForecastList<T> = T & WeatherInfo;

// Generic forecast interface
interface Forecast<T> {
  cod: string;
  message: number;
  cnt: number;
  city: CityInfo;
  list: ForecastList<T>[];
}

interface HourlyData {
  main: Record<string, number>;
  wind: Record<string, number>;
  visibility: number;
  rain: Record<string, number>;
  sys: Record<string, string>;
  dt_txt: string;
}

interface DailyData {
  sunrise: number;
  sunset: number;
  temp: Record<string, number>;
  feels_like: Record<string, number>;
  pressure: number;
  humidity: number;
  speed: number;
  deg: number;
  gust: number;
  rain: number;
}

// Reusable interfaces
type HourlyForecast = Forecast<HourlyData>;
type FiveDayForecast = Forecast<DailyData>;

type AQI = 1 | 2 | 3 | 4 | 5;
type AQIVerbose = "Good" | "Fair" | "Moderate" | "Poor" | "Very Poor";
type AirPollutionFailproof = AirPollution | "-";
interface AirPollution {
  coord: number[];
  list: {
    dt: number;
    main: {
      aqi: AQI;
    };
    components: Record<string, number>;
  }[];
}

interface CurrentForecast {
  coord: {
    lon: number;
    lat: number;
  };
  weather: Record<string, string | number>[];
  base: string;
  main: Record<string, number>;
  visibility: number;
  wind: Record<string, number>;
  rain: {
    "1h": number;
  };
  clouds: {
    all: number;
  };
  dt: number;
  sys: Record<string, string | number>;
  timezone: number;
  id: number;
  name: string;
  cod: number;
}

type WeatherError = {
  cod: string | number;
  message: string;
};

interface WeatherData {
  hourlyForecast: HourlyForecast | WeatherError;
  dailyForecast: FiveDayForecast | WeatherError;
  airPollution: AirPollutionFailproof | WeatherError;
  currentWeather: CurrentForecast | WeatherError;
}

type Time = `${number}:${number}`;
type Image = `${string}.svg` | `${string}.png` | `${string}.jpg`;
type Weekday =
  | "Today"
  | "Monday"
  | "Tuesday"
  | "Wednesday"
  | "Thursday"
  | "Friday"
  | "Saturday"
  | "Sunday";

interface WeatherContextType {
  weatherData: WeatherData | null;
  setWeatherData: React.Dispatch<React.SetStateAction<WeatherData | null>>;
  userInput: boolean;
  setUserInput: React.Dispatch<React.SetStateAction<boolean>>;
  currentCity: string;
  setCurrentCity: React.Dispatch<React.SetStateAction<string>>;
}

export type {
  WeatherData,
  CityCoordinates,
  HourlyForecast,
  FiveDayForecast,
  AQI,
  AQIVerbose,
  AirPollutionFailproof,
  CurrentForecast,
  Time,
  Image,
  Weekday,
  WeatherContextType,
  WeatherError,
};

export const WeatherContext = createContext<WeatherContextType | null>(null);
