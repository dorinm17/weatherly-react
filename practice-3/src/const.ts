// Since I'm not using a server, I can't store sensitive data like API keys in a .env file.
const OPENWEATHER_API_KEY: string = "c2b684a41108eb410abf5520de64edd6";
const GOOGLE_MAPS_API_KEY: string = "AIzaSyD4kq9y5crf0SYPJq-YxEhUtETDm5zfOBs";

enum sessionKeys {
  userInput = "userInput",
  currentCity = "currentCity",
  weatherData = "weatherData",
}

enum defaultCity {
  city = "Bucharest",
}

enum booleanStr {
  true = "true",
  false = "false",
}

export {
  OPENWEATHER_API_KEY,
  GOOGLE_MAPS_API_KEY,
  sessionKeys,
  defaultCity,
  booleanStr,
};
