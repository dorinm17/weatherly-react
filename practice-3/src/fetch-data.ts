import {
  GOOGLE_MAPS_API_KEY,
  OPENWEATHER_API_KEY,
  sessionKeys,
} from "./const.js";
import {
  AirPollutionFailproof,
  CityCoordinates,
  CurrentWeather,
  DailyForecast,
  HourlyForecast,
  WeatherData,
} from "./types.js";

// Use Google Maps Geocoding API to get the city name based on the user's coordinates, since OpenWeather is not highly precise.
const reverseGeocode = async (
  latitude: number,
  longitude: number
): Promise<string | null> => {
  const url: string = `https://maps.googleapis.com/maps/api/geocode/json?latlng=${latitude},${longitude}&key=${GOOGLE_MAPS_API_KEY}`;

  const city: string | null = await fetch(url)
    .then((response) => response.json())
    .then((data) => {
      if (!data.results || data.results.length === 0) {
        throw new Error("No results found for the given coordinates.");
      }

      const addressComponents = data.results[0].address_components;
      const localityComponent = addressComponents.find(
        (component: { types: string[]; long_name: string }) =>
          component.types.includes("locality")
      );

      return localityComponent ? localityComponent.long_name : null;
    })
    .catch((error) => {
      console.error("Error getting city name:", error);
      return sessionStorage.getItem(sessionKeys.currentCity);
    });

  return city;
};

const getCityCoordinates = async <CityCoordinates>(
  city: string
): Promise<CityCoordinates | null> => {
  const url: string = `https://api.openweathermap.org/geo/1.0/direct?q=${city}&limit=1&appid=${OPENWEATHER_API_KEY}`;

  const response: CityCoordinates[] = await fetch(url)
    .then((res) => res.json())
    .catch((error) => {
      console.error("Error fetching city coordinates:", error);
      return [];
    });

  // Return the first city coordinate if it exists, otherwise null
  return response.length > 0 ? response[0] : null;
};

// Use OpenWeatherMap API to retrieve the weather data for a given city and country code.
const getWeather = async (city: string): Promise<WeatherData> => {
  const coords: CityCoordinates | null = await getCityCoordinates(city);

  if (!coords) {
    throw new Error(`Coordinates not found for city: ${city}`);
  }

  const urls: string[] = [
    `https://pro.openweathermap.org/data/2.5/forecast/hourly?q=${city}&cnt=10&appid=${OPENWEATHER_API_KEY}&units=metric`,
    `https://api.openweathermap.org/data/2.5/forecast/daily?q=${city}&cnt=5&appid=${OPENWEATHER_API_KEY}&units=metric`,
    `https://api.openweathermap.org/data/2.5/air_pollution?lat=${coords.lat}&lon=${coords.lon}&appid=${OPENWEATHER_API_KEY}`,
    `https://api.openweathermap.org/data/2.5/weather?q=${city}&appid=${OPENWEATHER_API_KEY}&units=metric`,
  ];

  try {
    const responses = await Promise.all(
      urls.map((url, index) =>
        fetch(url)
          .then((response) => response.json())
          .catch((error) => {
            if (index === 2) {
              // Air Pollution request
              console.warn("Air pollution data not available. Setting as '-'.");
              return "-"; // Return placeholder value for airPollution
            }
            throw error; // Propagate other errors
          })
      )
    );

    // Ensure each response is properly typed as per the WeatherData interface
    const weatherData: WeatherData = {
      hourlyForecast: responses[0] as HourlyForecast,
      dailyForecast: responses[1] as DailyForecast,
      airPollution: responses[2] as AirPollutionFailproof,
      currentWeather: responses[3] as CurrentWeather,
    };

    return weatherData;
  } catch (error) {
    console.error("Error fetching weather data:", error);
    throw new Error("Failed to fetch weather data.");
  }
};

export { getWeather, reverseGeocode };
