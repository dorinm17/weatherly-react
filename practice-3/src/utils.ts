/**
 * @param {string} s
 */
const capitalize = (s) => {
  return s.charAt(0).toUpperCase() + s.slice(1);
};

/**
 * 
 * @param {number} sunrise 
 * @param {number} sunset 
 * @returns 
 */
const dayOrNight = (sunrise, sunset) => {
  const now = new Date().getTime() / 1000;
  return now >= sunrise && now < sunset;
};

// Display an accurate weather icon.
/**
 * 
 * @param {number} code 
 * @param {boolean} daytime 
 * @returns 
 */
const chooseImage = (code, daytime) => {
  let image;

  if (code >= 200 && code < 300) image = "thunderstorm";
  else if (code >= 300 && code < 600) image = "rain";
  else if (code >= 600 && code <= 613) image = "snow";
  else if (code >= 615 && code < 700) image = "rain-and-snow";
  else if ([701, 721, 741, 771].includes(code)) image = "wind";
  else if (code === 781) image = "tornado";
  else if (code >= 700 && code < 800) image = "smoke";
  else if (code === 800) image = daytime ? "sun" : "moon";
  else if ([801, 802].includes(code))
    image = daytime ? "sun-and-clouds" : "moon-and-clouds";
  else if (code >= 803) image = "clouds";

  const imageTag = document.createElement("img");
  imageTag.src = `images/${image}.svg`;
  imageTag.alt = "";
  return imageTag;
};

/**
 * 
 * @param {number} mps 
 * @returns 
 */
const convertMpsToKmph = (mps) => {
  return (mps * 3.6).toFixed(0);
};

// Convert the Air Quality Index to a more verbose description.
/**
 * 
 * @param {number} aqi 
 * @returns 
 */
const convertAQI = (aqi) => {
  const descArr = ["Good", "Fair", "Moderate", "Poor", "Very Poor"];
  return descArr[aqi - 1];
};

/**
 * 
 * @param {number} time 
 * @param {number} timezone 
 * @returns 
 */
const convertToLocalTime = (time, timezone) => {
  return new Date((time + timezone) * 1000).toLocaleTimeString("en-US", {
    timeZone: "UTC",
    hour: "numeric",
    minute: "2-digit",
    hour12: false,
  });
};

// Check if the temperature is zero and return 0 instead of -0.
/**
 * @param {number} temp
 */
const checkIfZeroTemp = (temp) => {
  const roundedTemp = temp.toFixed(0);
  return roundedTemp === "-0" ? 0 : roundedTemp;
}

export { capitalize, dayOrNight, chooseImage, convertMpsToKmph, convertAQI, convertToLocalTime, checkIfZeroTemp };
