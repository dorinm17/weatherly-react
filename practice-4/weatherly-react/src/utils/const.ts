// Since I'm not using a server, I can't store sensitive data like API keys in a .env file.
const OPENWEATHER_API_KEY: string = import.meta.env.VITE_OPENWEATHER_API_KEY;
const GOOGLE_MAPS_API_KEY: string = import.meta.env.VITE_GOOGLE_MAPS_API_KEY;

const defaultCity: string = "Bucharest";

export { OPENWEATHER_API_KEY, GOOGLE_MAPS_API_KEY, defaultCity };
