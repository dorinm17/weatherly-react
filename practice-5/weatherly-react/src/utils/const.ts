const OPENWEATHER_API_KEY: string = "c2b684a41108eb410abf5520de64edd6";
const GOOGLE_MAPS_API_KEY: string = "AIzaSyD4kq9y5crf0SYPJq-YxEhUtETDm5zfOBs";

const defaultCity: string = "Bucharest";

export { OPENWEATHER_API_KEY, GOOGLE_MAPS_API_KEY, defaultCity };
