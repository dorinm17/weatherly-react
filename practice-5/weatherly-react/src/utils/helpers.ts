import type { AQI, AQIVerbose, Time, Weekday, Image } from "./types";

// Decorator to check if the `number` type arguments for most of the methods are positive.
const isPositive = (paramIndexes?: number[]): MethodDecorator => {
  return function (
    _,
    propertyKey: string | symbol,
    descriptor: PropertyDescriptor
  ): void {
    const originalMethod = descriptor.value;

    descriptor.value = function (...args: number[]) {
      const indexesToCheck = paramIndexes ?? args.map((_, index) => index);

      indexesToCheck.forEach((index) => {
        if (typeof args[index] === "number" && args[index] < 0) {
          throw new Error(
            `Argument at index ${index} in ${String(
              propertyKey
            )} must be a non-negative number.`
          );
        }
      });

      return originalMethod.apply(this, args);
    };
  };
};

class Utils {
  static capitalize(s: string): string {
    return s.charAt(0).toUpperCase() + s.slice(1);
  }

  @isPositive()
  static dayOrNight(sunrise: number, sunset: number): boolean {
    const now: number = new Date().getTime() / 1000;
    return now >= sunrise && now < sunset;
  }

  // Display an accurate weather icon.
  @isPositive()
  static chooseImage(code: number, daytime: boolean): Image {
    let image: string = "";

    if (code >= 200 && code < 300) image = "thunderstorm";
    else if (code >= 300 && code < 600) image = "rain";
    else if (code >= 600 && code <= 613) image = "snow";
    else if (code >= 615 && code < 700) image = "rain-and-snow";
    else if ([701, 721, 741, 771].includes(code)) image = "wind";
    else if (code === 781) image = "tornado";
    else if (code >= 700 && code < 800) image = "smoke";
    else if (code === 800) image = daytime ? "sun" : "moon";
    else if ([801, 802].includes(code))
      image = daytime ? "sun-and-clouds" : "moon-and-clouds";
    else if ([803, 804].includes(code)) image = "clouds";

    const imageTag: HTMLImageElement = document.createElement("img");
    const imageURL: Image = `/src/assets/${image}.svg`;
    try {
      imageTag.src = imageURL;
      imageTag.alt = "";
    } catch (error) {
      console.error("Error loading image:", error);
    }

    return imageURL;
  }

  @isPositive()
  static convertMpsToKmph(mps: number): string {
    return (mps * 3.6).toFixed(0);
  }

  // Convert the Air Quality Index to a more verbose description.
  static descArr: AQIVerbose[] = [
    "Good",
    "Fair",
    "Moderate",
    "Poor",
    "Very Poor",
  ];
  static convertAQI(aqi: AQI): AQIVerbose {
    return Utils.descArr[aqi - 1];
  }

  @isPositive([0])
  static convertToLocalTime(time: number, timezone: number): Time {
    return new Date((time + timezone) * 1000).toLocaleTimeString("en-US", {
      timeZone: "UTC",
      hour: "numeric",
      minute: "2-digit",
      hour12: false,
    }) as Time;
  }

  // Check if the temperature is zero and return 0 instead of -0. Return the rounded temperature in any case.
  static checkIfZeroTemp(temp: number): number {
    const roundedTemp: string = temp.toFixed(0);
    return parseFloat(roundedTemp === "-0" ? "0" : roundedTemp);
  }

  @isPositive()
  static getWeekday(timestamp: number): Weekday {
    return new Date(timestamp * 1000).toLocaleDateString("en-US", {
      weekday: "long",
    }) as Weekday;
  }
}

export const capitalize = Utils.capitalize;
export const dayOrNight = Utils.dayOrNight;
export const chooseImage = Utils.chooseImage;
export const convertMpsToKmph = Utils.convertMpsToKmph;
export const convertAQI = Utils.convertAQI;
export const convertToLocalTime = Utils.convertToLocalTime;
export const checkIfZeroTemp = Utils.checkIfZeroTemp;
export const getWeekday = Utils.getWeekday;
