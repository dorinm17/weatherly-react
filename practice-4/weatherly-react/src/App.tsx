import Header from "./components/Header/Header";
import LoadingSpinner from "./components/LoadingSpinner/LoadingSpinner";
import Footer from "./components/Footer/Footer";
import DailyForecast from "./components/DailyForecast/DailyForecast";
import CurrentWeather from "./components/CurrentWeather/CurrentWeather";
import TodayForecast from "./components/TodayForecast/TodayForecast";
import { useEffect, useState, useMemo } from "react";
import { WeatherContext, WeatherData, WeatherContextType } from "./utils/types";
import { defaultCity } from "./utils/const";
import { getWeather, reverseGeocode } from "./utils/fetch-data";

function App() {
  const [userInput, setUserInput] = useState<boolean>(false);
  const [currentCity, setCurrentCity] = useState<string>(defaultCity);
  const [weatherData, setWeatherData] = useState<WeatherData | null>(null);
  const [loading, setLoading] = useState(true);

  // When the current session ends, reset the weather to the user's current location.
  useEffect(() => {
    const controller: AbortController = new AbortController();

    const handleBeforeUnload = () => {
      setUserInput(false);
      setCurrentCity(defaultCity);
      setWeatherData(null);
      setLoading(true);
    };

    window.addEventListener("beforeunload", handleBeforeUnload, {
      signal: controller.signal,
    });

    // Cleans up all event listeners associated with `controller.signal`
    return () => controller.abort();
  }, []);

  // Before the user gets to search for a city, the app will display the weather for the default city.
  useEffect(() => {
    let city: string = defaultCity;

    (async () => {
      try {
        if (navigator.geolocation) {
          // Wrap geolocation in a Promise to ensure we wait for it.
          const position: GeolocationPosition =
            await new Promise<GeolocationPosition>((resolve, reject) => {
              navigator.geolocation.getCurrentPosition(resolve, reject);
            });
          const latitude: number = position.coords.latitude;
          const longitude: number = position.coords.longitude;

          city = (await reverseGeocode(latitude, longitude)) ?? defaultCity;
          setCurrentCity(city);
        } else console.log("Geolocation is not supported by this browser.");
        // Now that we have the actual current location, fetch the weather
        const data: WeatherData = await getWeather(city);
        setWeatherData(data);
        setLoading(false);
      } catch (error) {
        if ((error as GeolocationPositionError).code === 1) {
          console.error("User denied geolocation access.");
          setLoading(false);
          return;
        }
        console.error("Error in initial weather process:", error);
      }
    })();
  }, []);

  // Memoize the context value to avoid unnecessary re-renders
  const contextValue = useMemo<WeatherContextType>(
    () => ({
      weatherData,
      setWeatherData,
      userInput,
      setUserInput,
      currentCity,
      setCurrentCity,
    }),
    [weatherData, userInput, currentCity]
  );

  return (
    <WeatherContext.Provider value={contextValue}>
      <Header />

      {loading && <LoadingSpinner />}
      {!loading && (
        <main>
          <DailyForecast />
          <TodayForecast />
          <CurrentWeather />
        </main>
      )}

      <Footer />
    </WeatherContext.Provider>
  );
}

export default App;
