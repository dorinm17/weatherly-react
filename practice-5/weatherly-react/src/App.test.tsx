import { render, screen, act } from "@testing-library/react";
import App from "./App";
import { WeatherData } from "./utils/types";
import { getWeather, reverseGeocode } from "./utils/fetch-data";
import { defaultCity } from "./utils/const";

jest.mock("./utils/fetch-data", () => ({
  getWeather: jest.fn(),
  reverseGeocode: jest.fn(),
}));

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useEffect: jest.requireActual("react").useEffect,
  useState: jest.requireActual("react").useState,
  useMemo: jest.requireActual("react").useMemo,
}));

describe("App Component", () => {
  beforeEach(() => {
    (getWeather as jest.Mock).mockResolvedValue({
      currentWeather: { cod: 200 },
      dailyForecast: {},
      hourlyForecast: {},
      airPollution: {},
    } as WeatherData);

    (reverseGeocode as jest.Mock).mockResolvedValue("Mock City");

    const mockGeolocation = {
      getCurrentPosition: jest.fn().mockImplementation((success) =>
        success({
          coords: {
            latitude: 51.5074,
            longitude: -0.1278,
          },
        })
      ),
      clearWatch: jest.fn(),
      watchPosition: jest.fn(),
    };

    // Mock geolocation using Object.defineProperty
    Object.defineProperty(global.navigator, "geolocation", {
      value: mockGeolocation,
      writable: true,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders loading spinner initially", () => {
    render(<App />);
    expect(screen.getByTestId("loading-spinner")).toBeInTheDocument();
  });

  test("successfully geolocates the user", async () => {
    await act(async () => {
      render(<App />);
    });

    expect(getWeather).toHaveBeenCalledWith("Mock City");
    expect(screen.queryByTestId("loading-spinner")).not.toBeInTheDocument();
  });

  test("handles geolocation fetch error gracefully", async () => {
    (
      navigator.geolocation.getCurrentPosition as jest.Mock
    ).mockImplementationOnce((_, reject) => reject({ code: 2 }));

    await act(async () => {
      render(<App />);
    });

    expect(screen.queryByTestId("loading-spinner")).not.toBeInTheDocument();
  });

  test("handles denial of geolocation by user gracefully and fetches weather for default city", async () => {
    (
      navigator.geolocation.getCurrentPosition as jest.Mock
    ).mockImplementationOnce((_, reject) => reject({ code: 1 }));

    await act(async () => {
      render(<App />);
    });

    expect(getWeather).toHaveBeenCalledWith(defaultCity);
    expect(screen.queryByTestId("loading-spinner")).not.toBeInTheDocument();
  });

    test("matches snapshot", () => {
      const { asFragment } = render(<App />);
      expect(asFragment()).toMatchSnapshot();
    });
});
