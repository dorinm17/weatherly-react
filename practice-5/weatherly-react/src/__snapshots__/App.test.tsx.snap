// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches snapshot 1`] = `
<DocumentFragment>
  <header>
    <div>
      <h1>
        Weatherly
      </h1>
      <h2>
        Don't be under the weather
        <span>
          !
        </span>
      </h2>
    </div>
    <img
      alt=""
      data-testid="banner-image"
    />
    <nav>
      <form
        action=""
        aria-label="City search form"
        autocomplete="off"
        method="get"
      >
        <button
          aria-label="Submit your search"
          type="submit"
        >
          <img
            alt="search"
          />
        </button>
        <input
          aria-label="Search for your city"
          name="city"
          placeholder="Search for your city"
          type="text"
          value=""
        />
      </form>
    </nav>
  </header>
  <div
    data-testid="loading-spinner"
  >
    <div />
  </div>
  <footer>
    <p>
      Â© 2025 Weatherly, LLC. All rights reserved.
    </p>
    <ul
      aria-label="Social Media Links"
      data-testid="links-list"
    >
      <li>
        <a
          href="https://www.linkedin.com/in/dorin-manea/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="LinkedIn"
          />
        </a>
      </li>
      <li>
        <a
          href="https://www.instagram.com/dorinm17/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="Instagram"
          />
        </a>
      </li>
      <li>
        <a
          href="https://www.facebook.com/dorinm17/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="Facebook"
          />
        </a>
      </li>
      <li>
        <a
          href="mailto:dmanea@adobe.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="Mail"
          />
        </a>
      </li>
    </ul>
  </footer>
</DocumentFragment>
`;
